\documentclass[a4paper,10pt]{article}
\usepackage{a4}
\usepackage{epsfig}
\parindent=0pt
\parskip=10pt
%\renewcommand{\baselinestretch}{2}

\title{Running the Ensembl Anopheles automatic annotation} 
\author{Emmanuel Mongin (mongin@ebi.ac.uk)}

\begin{document}

\maketitle


\bf{File last updated: August 2003}
The goal is to provide a written help on how to run the the \textit{A. gambiae} automatic annotation. This should actually be usable for most of the 'virgin genomes'. This documentation for processes which are common with other genomes, references are given to files present in the ensembl CVS repository called ensembl-doc. These files should be updated as the processes and code change. All of this documentation is written for people working at EBI/Sanger. Please update this documentation if needed. 

\section{Raw computes}
Running raw computes for \textit{A. gambiae} is pretty similar than for other genomes only few details change.
The document available is called using\_the\_ensembl\_pipeline. This is a pretty extensive and detailed document.

Here are the raw computes which have been run for the release 3:
\begin{itemize}
\item RepeatMask
\item TRF
\item tRNAscan
\item CPG island
\item DUST
\item Dros pep blastX hits
\item Swall blastX hits
\item Snap genes
\item Genscan (using the arabidopsis matrix)
\end{itemize}

\textbf{Particularities}\\
\textbf{RepeatMasker} is ran with a customed library (which need to be placed on the following directory: XXXX). The parameter used should be the following: \--lib anopheles.\\
\textbf{Genscan} is not worth to be reran. This gave pretty bad gene structure. Does not bring any useful information.\\
\textbf{Snap} (another \textit{ab initio} gene predictor) trained with anopheles specific ESTs (see section on how to train SNAP) gives much better result. In many regions it seems to get the right exon prediction, it does not tend to overpredict too much. However the transcript structure is not always right. Snap is used in the last part of the gene build and some SNAP genes (about 2000) will be elevated to the status of ***REMOVED*** gene.\\
\textbf{BlastX} is ran with two different datasets. The first one (and maybe the most important one) is the protein dataset containing all of the sequences from flybase. The second one is Swall.\\

NB: Make sure that the logic names in the analysis table. Please check with the web team.\\
The whole pipeline process (setting up the database and running it) should take about 3 days. On of the thing to do to get the analysis and rule tables right is to copy the ones from the previous release and have a critical look at them.

\section{Gene Build}
Gene build is perhaps the most tricky part of the process. Here are some of the reasons:
\begin{itemize}
\item No close annotated organism
\item Really few genes annotated (Immunity, odorant receptors, ...)
\item Few ESTs/cDNAs available (190 000 at 08/03)
\end{itemize}
One of the main change made to the process is to use much lower score blast hits. But this has many effects:

Here are the main issues, genebuilding on \textit{An. gambiae}:
\begin{itemize}
\item Extremely high number of potential parent proteins to be used fo the similarity process. This has for consequence an increase of the compute time and more noise brought to the exon structure
\item Genewise may built incomplete exons on low score blast hits. Transcritpts build with incomplete exons may not be merged with the correct exon structure and then produce wrong splice variants.
\item Problems in merging splice variant from a same gene. This is due to partial exon structure build through the similarity process (mainly due to the use of low score blast hits). As EST genes are being merged to the final gene build. These genes in some case are not merged properly with similarity genes
\item Incomplete Repeat dataset
\item False negative. Still a certain number (about 10\% for the last release) may be missed
\end{itemize}

\subsection{Gene build by similarity}
The score of the blast hits used to build similarity gene is much lower than for other organisms. We used a score down to 125 (lower than that seems to only bring noise, higer should obviously reduce the noise but some edges exons may be missed).
\\
NB: Blast hit filter has absolutely to be turned on whereas the similarity jobs would take too long or never finish.

\subsubsection{Targetted}
Taregetted process is undertaken with 3 dataset:
\begin{enumerate}
\item \textit{An. gambiae} genes in Swiss-Prot and SPTrEMBL.
\item Submitted genes to the gene name submission database. (a script exists in the script section of the anopheles-genename cvs repository). If a cDNA has been submitted, the script will dump the longuest open reading frame.
%get the name of the script
\item Protein available for the latest \textit{D. melanogaster} genome. (usually dump them from the Ensembl version of Drosophila. This will allow to predict the drosophila orthologs.
\end{enumerate}

Here is an example of the conf file used for the last release (release 3).%add conf file here
\begin{verbatim}
  # minimum required coverage for multiexon predictions
  GB_TARGETTED_MULTI_EXON_COVERAGE      => '25',
  
  # minimum required coverage for single predictions
  GB_TARGETTED_SINGLE_EXON_COVERAGE     => '90',
  
  # maximum allowed size of intron in Targetted gene
  GB_TARGETTED_MAX_INTRON               => '50000',
  
  # minimum coverage required to prevent splitting 
  #on long introns - keep it high!
  GB_TARGETTED_MIN_SPLIT_COVERAGE       => '95',
  
  # genetype for Targetted_GeneWise
  GB_TARGETTED_GW_GENETYPE              => 'TGE_gw',	   
  GB_TARGETTED_MASKING => ['RepeatMask', 'Dust'],
  #the above setting will lead no features being 
  #masked out, if is used no masking will take 
  #place, if [''] is used all repeats in the 
  #table will be masked out
  #if only some of the sets of repeats wants 
  #to be masked out put logic names of the 
  #analyses in the array 
  #e.g ['RepeatMask', 'Dust']
  GB_TARGETTED_SOFTMASK => 0, 
  # set this to one if you want the repeats 
  #softmasked ie lower case rather 
  than uppercase N's
\end{verbatim}

\textbf{potential problems}:
\begin{itemize}
\item Check for XXXs in the protein sequence. If a sequence has too many XXXs, pmatch may never end. Generally check that pmatch properly finished, it would fail if there is bad characters. Some of the drosophila proteins definitely have to be deleted.
\item After running pmatch(see ensembl-doc documentation) check at this step if most of the proteins have been mapped to the genome. 
\end{itemize}

\subsubsection{Similarity}
The similarity process uses both blast hits from Swall and Drosophila blastX hits. With the blast filtering code on (see example of the conf file) all of the jobs should be finished after 24h (on the farm if not too loaded).
\\
Here si an example of hte configuration file used for the last anopheles release for the similarity process:

\begin{verbatim}

  GB_SIMILARITY_DATABASES => [
    {				  
      'type'       => 'Swall',
      'threshold'  => '125',
      'index'      => '/data/blastdb/Ensembl/swall_030604',
      'seqfetcher' => 'Bio::EnsEMBL::Pipeline::
       SeqFetcher::OBDAIndexSeqFetcher'
    },
    {
      'type' => 'drosphila-peptides',
      'threshold' => '125',
      'index'   => 'Ensembl/drosophila-release3-peptides-unique_index',
      'seqfetcher' => 'Bio::EnsEMBL::Pipeline::
       SeqFetcher::OBDAIndexSeqFetcher',
    }
  ],
  # minimum required parent protein coverage
  GB_SIMILARITY_COVERAGE           => 40,
  
  # maximum allowed size of intron 
  GB_SIMILARITY_MAX_INTRON         => 10000,
  
  # minimum coverage required to prevent splitting on long introns - 
  #keep it high!
  GB_SIMILARITY_MIN_SPLIT_COVERAGE => 90,
  
  # low complexity threshold - transcripts whose translations have low
  # complexity
  GB_SIMILARITY_MAX_LOW_COMPLEXITY => 70,
  
  # gene type for FPC_BlastMiniGenewise
  
  GB_SIMILARITY_GENETYPE           => 'similarity_pruned1',
  
  GB_SIMILARITY_MASKING => ['RepeatMask','DUST'],
  
  #the above setting will lead to RepeatMasker features being masked out, 
  #if  used no masking will take place, if [''] is used all repeats in the 
  #table will be masked out
  #if only some of the sets of repeats wants 
  #to be masked out put logic names 
  #of the analyses in the array e.g ['RepeatMask', 'Dust']
  GB_SIMILARITY_SOFTMASK => 0, 
  
  # set this to one if you want the repeats 
  softmasked ie lower case rather than uppercase N's
  #No similarity genes will be build overlapping the 
  #gene type put in this list. If nothing is put, 
  #the default will be to take 
  #Targetted genewise genes
  
  GB_SIMILARITY_GENETYPEMASKED => ['TGE_gw','genomewise_final'],
  
  GB_SIMILARITY_BLAST_FILTER => 1,
  
\end{verbatim}


\textbf{potential problems}: 
\begin{itemize}
\item The score used for this run are low (125 for the blast hits) in some locus many spliced variants (or what will become spliced variants) with incomplete exons. Some of the transcripts may also span on two transcripts locus.
\item In the last gene build there is few cases where genes should have been build with similarity and where nothing was build. I'll leave a list of location where genes have been missed.
\item As we use low score blast hits a lot of wrong spliced variants may be build in later stage. One of the trick I've been using (that's definitely not the solution but this help waiting to find a solution), is to run the gene build over the similarity genes and limit the number of transcripts allowed for one cluster to 3. In that case only the 3 best transcripts from a cluster will be use.

\end{itemize}

\subsubsection{Combine with cDNAs}
This process will combine the cDNAs to the similarity genes to add UTRs where possible. The process used here is the same than for human. The only difference is (see EST gene build section) that we don't do separate EST gene build and cDNA gene build, thus all of the data available (EST + cDNA mapped) are used for this step.

\subsection{EST Gene build}
\subsubsection{Exonerate and the EST gene builder}
\paragraph{Exonerate}
The EST gene build is ran without much adaptation to the process (for more detaisl see the conf files). As the number of ESTs anopheles remains relatively low (less than 200 000) and that they are of pretty good quality, we treat these data as cDNAs. For the last release all of the \textit{An. gambiae} sequences labelled as RNA have been used for the gene build. 
To map the EST to the genome, only exonerate has been used.
For documentation on how to run this build read in ensemb-doc, cdna analysis.txt.
Here is the configuration file used to run the Exonerate step (mapping of the ESTs to the genome):

\begin{verbatim}
  EST_INPUTID_REGEX => '(\S+)\.(\d+)-(\d+)',
  EST_RUNNER        => 'pipeline-new/
                        scripts/EST/run_exonerate.pl', 	   
  # path to pipeline-new/scripts/EST
  EST_SCRIPTDIR     => 'pipeline-new/scripts/EST/',
	      
  # where the result-directories are going to go	
  EST_TMPDIR        => 'est_build/',	      
  # job-queue in the farm
  LSF_OPTIONS       => '-q acari -C0',
  
  # make_bsubs.pl options
  EST_EXONERATE_BSUBS   => 'ests/bsubs/exonerate.bsub',
  
  # path to file containing ALL ESTs/cDNAs
  EST_FILE                    => 'ests/total_ests_plus_submitted.fa',
  
  # path to directory where EST chunks live
  EST_CHUNKDIR                => 'ests/est_chunks/',
  
  # how many chunks?
  # for NCBI_28 we have 3690891 ests, at approx. 
  #350 ests per chunk, we estimate
  EST_CHUNKNUMBER             => 550, 	 
  	      
  # full path fo the dir where we have 
  #the masked-dusted chromosomes
  EST_GENOMIC                 => 'mosquito/genome/',
  
  # path to file with repeatmasked dusted genomic sequences
  # NB this file is huge - distribute it across the farm or 
  # be prepared to face the wrath of systems when the network 
  # seizes up!
  
  EST_EXONERATE              => 'ensembl/bin/exonerate-0.6.7',
  
  # if set to true, this option rejects unspliced alignments 
  #for cdnas that have an spliced
  # alignment elsewhere in the genome
  REJECT_POTENTIAL_PROCESSED_PSEUDOS => 0,
  
  # if set to true, the only the best match 
  #in the genome is picked
  # if there are several matches with the same coverage
  # all of them are taken, except single-exon 
  # ones if REJECT_POTENTIAL_PROCESSED_PSEUDOS is switched on 
  BEST_IN_GENOME => 1,
  
  EST_MIN_COVERAGE            => 95,
  EST_MIN_PERCENT_ID          => 97,
  

  EST_USE_DENORM_GENES	=> 0,
  
  ############################################################
  # each runnable has an analysis
  ############################################################
  
  EST_EXONERATE_RUNNABLE     => 'Bio::EnsEMBL::Pipeline::
   RunnableDB::ExonerateToGenes',      
  EST_EXONERATE_ANALYSIS     => 'RNA_BEST',
  
  EST_SOURCE                  => 'EMBL',      
  
\end{verbatim}

\paragraph{EST Gene builder}
The configuration file for the EST gene build is pretty complex. The best is to chat with eduardo (eae@sanger.ac.uk) because there is a lot of development being undertaken and the following conf file may be outdated in a month 

\begin{verbatim}
  ############################################################
  # EST_GeneBuilder
  ############################################################
  
  EST_GENEBUILDER_CHUNKSIZE        => 1000000,      
  #  we use 1000000 (ie 1MB) chunks
  
  EST_GENEBUILDER_INPUT_GENETYPE => 'RNA',
  EST_GENOMEWISE_GENETYPE        => 'genomewise_final',
  
  # if this is set to TRUE it will reject ests that do not
  # have all splice sites correct
  CHECK_SPLICE_SITES => 1,
  
  # if set to a number, it will reject single exon ests 
  #that are shorter that this
  FILTER_ON_SINGLETON_SIZE => 500,
  
  # if set to a number, it will reject single exon ests 
  #that have score smaller than this
  RAISE_SINGLETON_COVERAGE => 97,
  
  ## you must choose one type of merging for cdnas/ests: 
  #2 and 3 are the common ones
  EST_GENEBUILDER_COMPARISON_LEVEL => 2,
  
  # for details see documentation 
  # in Bio::EnsEMBL::Pipeline::GeneComparison::TranscriptComparator
  # 1 --> strict: exact exon matching (unrealistic). 
  # 2 --> allow edge exon mismatches
  # 3 --> allow internal mismatches
  # 4---> allow intron mismatches
  # 5---> loose mode - consecutive exon overlap - allows intron mismatches
  
  # you can alow a mismatch in the splice sites
  EST_GENEBUILDER_SPLICE_MISMATCH  => 40,
  
  # you can allow matches over small introns 
  EST_GENEBUILDER_INTRON_MISMATCH => 20,
  
  # you can bridge over small introns: 
  #we difuse the small intron into one exon
  # if set to false transcripts with small introns will be rejected
  BRIDGE_OVER_SMALL_INTRONS => 0,
  
  # the maximum size of introns to bridge over
  EST_MIN_INTRON_SIZE  => 20,
  
  
  # you can choose whether you only want tw ests/cdnas to merge if
  # they have the same number of exons
  EST_GENEBUILDER_EXON_MATCH     => 0,
  
  # how much discontinuity we allow in the supporting evidence
  # this might be correlated with the 2-to-1 merges, so we
  # usually put it =  EST_GENEBUILDER_INTRON_MISMATCH for ESTs
  EST_MAX_EVIDENCE_DISCONTINUITY  => 2,
  REJECT_SINGLE_EXON_TRANSCRIPTS  => 0,
  GENOMEWISE_SMELL                => 0,
  
  # exons smaller than this will not be include in the merging algorithm
  EST_MIN_EXON_SIZE               => 10,
  
  # ests with intron bigger than this will not be incuded either
  EST_MAX_INTRON_SIZE             => 50000,
  
  # this says to ClusterMerge what's the minimum
  # number of ESTs/cDNAs that must be 'included' into a
  # final transcript
  CLUSTERMERGE_MIN_EVIDENCE_NUMBER => 1,
  
  # maximum number of transcripts allowed to 
  # be in a gene. Even by tuning the other parameteres
  # to keep this low, there will be always 
  #cases with more 20 even 50 isoforms
  # which, unless what you're doing is really targetted
  # to a known case or with very good quality ests/cdnas, it
  # is not very reliable.
  MAX_TRANSCRIPTS_PER_GENE => 8,
  
  # If using denormalised gene table, set this option to 1.
  EST_USE_DENORM_GENES => 0,
            
\end{verbatim}

\bf{NB:} One of the most important option of the configuration is the genebuilder comparison level. For the Anopheles EST gene build release 3 we used level 2. See the configuration file above or even better Eduardo's Cluster Merge paper to get a better idea of the comparison levels.

\subsubsection{GeneCombiner (merging EST genes with the similarity build)}
The gene combiner will combine EST genes with similarity genes. This will add EST genes where no similarity genes have been predicted and add spliced variants to similarity genes.
This is still under heavy developpment by Eduardo. Talk to him next time this process is ran. Many think may have changed. Looking at the release 3 data, it seems that some of the EST genes have not been added to the final set where no other similarity genes have been predicted or not added to a gene structure, where we would expect them to add a new splice variant. This issues have been sent to Eduardo who is looking into it.

\section{Adding SNAP genes}
SNAP is an \textit{ab initio} gene predictor written by Ian Korf (ik1@sanger.ac.uk). This program is trained with a specific set of Anopheles EST genes (see bellow) by Ian. This is so far the best open source gene predictor we have tested for anopheles. Snap genes will be added where neither similarity nor EST genes have been predicted. To add these genes we use the GeneBuilder over the set of genes coming from the GeneCombiner with a specific specific set of options. 

\section{Post Compute}
\subsection{Gene build check}
\subsection{Protein Annotation}
\subsection{ID mapping}
\subsection{Known gene mapping}
\subsection{Database integrity checks}

\section{misc}
\subsection{Produce a training dataset for SNAP}
\subsection{Mapping BAC clones}
\subsection{Mapping markers}
\subsection{Transosons}

\section{Future developpements}

\end{document}
